src/test/main.cpp

```diff
@@ -0,0 +1,1154 @@
+                    }
+                }
+
+                // Remove found section from list to avoid being report.
+                gSection = gen_sections.erase(gSection);
+                cache_sections.erase(foundSection);
+            }
+
+            // output which sections has not been found
+            local_error_count += gen_sections.size() + cache_sections.size();
+            for (const auto& gSection : gen_sections) {
+                XbSUT_OutputMessage<false>(XB_OUTPUT_MESSAGE_ERROR, std::string("cache file does not have [") + gSection.pItem + "] section!");
+                gen_result.GetAllKeys(gSection.pItem, gen_keys);
+                for (const auto& gKey : gen_keys) {
+                    gen_result.GetAllValues(gSection.pItem, gKey.pItem, gen_values);
+                    local_error_count += gen_values.size();
+                    for (const auto& gValue : gen_values) {
+                        XbSUT_OutputMessage<false>(XB_OUTPUT_MESSAGE_ERROR, std::string("cache file does not have [") + gSection.pItem + "] " + gKey.pItem + " = \"" + gValue.pItem + "\" value!");
+                    }
+                }
+            }
+            for (const auto& cSection : cache_sections) {
+                XbSUT_OutputMessage<false>(XB_OUTPUT_MESSAGE_ERROR, std::string("generated result does not have [") + cSection.pItem + "] section!");
+                cache_result.GetAllKeys(cSection.pItem, cache_keys);
+                for (const auto& cKey : cache_keys) {
+                    cache_result.GetAllValues(cSection.pItem, cKey.pItem, cache_values);
+                    local_error_count += cache_values.size();
+                    for (const auto& cValue : cache_values) {
+                        XbSUT_OutputMessage<false>(XB_OUTPUT_MESSAGE_ERROR, std::string("generated result does not have [") + cSection.pItem + "] " + cKey.pItem + " = \"" + cValue.pItem + "\" value!");
+                    }
+                }
+            }
+
+            if (local_error_count) {
+                XbSUT_OutputMessage<false>(XB_OUTPUT_MESSAGE_ERROR, "XbSymbolUnitTest: FAIL - " + std::to_string(local_error_count) + " errors; generated result does not match with cache file!");
+                error_count++;
+            }
+
+            // Check force overwrite argument exist
+            if (!cli_config::hasKey("f")) {
+                doSave = false;
+            }
+        }
+        // Check if we want save to cache file.
+        if (doSave) {
+            gen_result.SaveFile(cache_file.c_str(), false);
+        }
+    }
+
+    if (full_lib_count) {
+        XbSUT_OutputMessage(XB_OUTPUT_MESSAGE_INFO, "XbSymbolUnitTest: Total of " + std::to_string(full_lib_count) + " full libraries found");
+    }
+
+    test_ret = output_result_XbSDb();
+
+    XbSUT_OutputMessage<false>(XB_OUTPUT_MESSAGE_INFO, "Unit test end.");
+
+    if (error_count) {
+        pause_for_user_input();
+    }
+
+    return test_ret;
+}
```

Comment:
Almost 1000 lines for this unit-test seems pretty brutal, with lots of redundancy and boilerplate. Is this just a modified copy of the original code?

Reply:
This unit test is intended to be brutal in order to reduce false positive report from contributors and testers. And ensure they are working as intended. The original code was from XbSymbolCacheGenTest (from Cxbx-Reloaded organization) which then dropped to develop XbSymbolUnitTest (from my personal repository). Finally, I made XbSymbolUnitTest more suitable to migrate into this XbSymbolDatabase repository. Mainly to maintenance databases remain in sync to any changes and find any faults/discovery from contributors' end.

On modern CPUs, it doesn't take long to finish one test. For multiple tests in one run for around 100 xbes, it only take maybe 30 seconds. This is for release build btw.

Reply:
I wasn't so much concerned for runtime duration, but more about code size and errors slipping in.
I feel like a large test harness adds more room for errors.

Reply:
I'm not sure how errors can slip in when I had made optimization to reduce redundant output to only one? Before, it used to be larger until I figured out a method to keep code smaller while have same message output only once with a counter included.

Source code aren't always final design and can have improvements any time in the future. Currently, it is stable enough to upstream. There will be at least couple areas to break down into functions in the future.

P.S. I had personally tested each block of tests, even with manual modification to cache file. Each has return working as intended. Otherwise, I wouldn't bother make this pull request.